### PR TITLE
Add application

### PR DESCRIPTION
**Pull request by diarisdiakite**
In this PR, I followed the [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including: Colors, Typography (font face, size and weight), Layout (composition and space between elements). I created:

📃 A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

📃 Sign up and log in pages
✔️ The user should be able to register in the app with full name, email and password (all mandatory).
✔️ The user can log into the app using email and password.
✔️ If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

📃 A Home page (categories page)
✔️ When the user logs in, they are presented with the categories page.
✔️ For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
✔️ When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
✔️ There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

📃 A Transactions page (Expenses page)
✔️ For a given category, the list of transactions is presented, ordered by the most recent.
✔️ At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
✔️ There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
✔️ When the user clicks on the "Back" button (<), the user navigates to the home page.

📃 An "Add a new category" page
✔️ The user fills out a form to create a new category, indicating their name and icon (both mandatory).
✔️ The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
✔️ When the user clicks on the "Back" button (<), the user navigates to the home page.

📃 An "Add a new transaction" page where the user fills out a form to create a new transaction with:
✔️ name (mandatory)
✔️ amount (mandatory)
✔️ categories (mandatory at least one)
✔️ The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
✔️ When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

📃 Fullfilled the Testing requirements
✔️ Create unit and integration tests for all the most important components of your RoR application.

📃 Technical requirements
✔️ Used Postgres as the project database.
✔️ Used devise for authentication.
✔️ Validate user inputs to make sure that anyone with bad intentions cannot compromise your app.
✔️ Used the view template engine '.erb'.
✔️ Deployed the project and made it accessible online.
✔️ Made sure my database schema should reflects the given structure:

ERD diagram

📃 For the Project documentation, I recorded a 5mns video presenting the features of the project. The content of the video includes:

✔️ a description of the project.
✔️ a demo of the project features.
✔️ some interesting piece of code I am very proud of highlighted

📃 I gave appropriate credit to the author in the README of the project as required by the [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+- Authentication with Devise
+- Authorization rules for users
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+## 🚀 Live demo <a name="live-demo"></a>
+
+Please watch the presentation video at this [link](https://www.loom.com/share/2747aeabe45443709317181533a60d6b)
+
```
Comment (Olamarx) [inline README.md]: - [x] While you did well recording, I noticed you laughed at the end of the video, it must be because the app has a bug, I also encountered the bug, I will encourage you to record another one that will be detailed, with a piece of code that interests you and complete the explanation of your project, you have tried with the short piece, just complete it. Remember to fix your bug too.

File: app/views/shared/_header_categories.html.erb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+          <%= link_to "Login", new_user_session_path, class:"button" %>
+          <%= link_to "Signup", new_user_registration_path, class:"button" %>
+        </li>
+      <% end %>
+  </ul>
+</nav>
+<div class="gift-credit-container details-heading">
+  <h2>Gift credit</h2>
+  <h2>$300</h2>
+</div>
```
Comment (Olamarx) [inline app/views/shared/_header_categories.html.erb]: - [x] I noticed you default some data upon login as I have circled, I will encourage you to empty it and leave all to the control of the user since it is their account.
image

File: app/views/shared/_footer.html.erb [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<div class="footer-container">
+  <%= link_to 'Add a New Category', new_user_category_path(@user), :class=>'footer-btn' %>
+</div>
```
Comment (Olamarx) [inline app/views/shared/_footer.html.erb:3]: - [x] When I click this button, the app breaks just as you also noticed when you were recording.
image

### Note for next reviewer
I could not confirm other requirements because the link that should get me there breaks the app. Kindly check it.

File: db/schema.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    t.string "checksum"
+    t.datetime "created_at", null: false
+    t.index ["key"], name: "index_active_storage_blobs_on_key", unique: true
+  end
+
+  create_table "active_storage_variant_records", force: :cascade do |t|
+    t.bigint "blob_id", null: false
+    t.string "variation_digest", null: false
+    t.index ["blob_id", "variation_digest"], name: "index_active_storage_variant_records_uniqueness", unique: true
+  end
```
Comment (Olamarx) [inline db/schema.rb]: - [x] I don't think you need all these tables for this project, you just need few tables which are below. However, in your categories table, which I think you replaced for group as it is in the requirement, you have name, created_at but there is no icon, kindly add the icon too 
image

**Review by Olamarx — CHANGES_REQUESTED** [review]
Hi @diarisdiakite, 

Good job so far, indeed you have spent time on this project!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
- Linters passed ✔️
- Readme file professional ✔️
- Commit messages meaningful ✔️

## Required Changes ♻️
_Check the comments under the review._
- [ ] I noticed you don't have the Splash Screen upon starting the app, kindly note that the splash screen should have only links to the sign up and log in pages with the name of your app and not the login page itself.
image
- [ ] Two of the tests are also not passing, even after precompiling the assets folder, kindly double-check it.
image

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: db/schema.rb (hunk 4 above)
Comment (diarisdiakite) [reply db/schema.rb]: The tables are the default tables of the 'activestorage' gem that I use for my images. They were added on the gem's installation. Active storage handles the icons that why I had to delete the column in my table.

**Comment by diarisdiakite** [comment]
Please find a screenshoot of the tests passing.

image

File: app/views/shared/_header_categories.html.erb (hunk 2 above)
Comment (diarisdiakite) [reply app/views/shared/_header_categories.html.erb]: Thank you for the comment @Olamarx . Maybe I have misunderstood but with the defaulted Gift credit, I tried to follow the design and for the house rent, this was a record I made to test the application. This can be removed from the category view page. Please find the design image for the gift credit.
image

File: README.md (hunk 1 above)
Comment (diarisdiakite) [reply README.md]: Hi @Olamarx, thank you for the comment, you are totally right and Sorry, I just copied the wrong link, as I fixed the bug before submitting. Please find the link to the right loom recording: https://www.loom.com/share/678be0ed9a374054bc2f6ba576425e89
This has been updated in the Readme file as well.

File: app/views/shared/_footer.html.erb (hunk 3 above)
Comment (diarisdiakite) [reply app/views/shared/_footer.html.erb:3]: Cheking what is happening in here. Thank you for the optional comment, I am fixing it.

File: README.md (hunk 1 above)
Comment (Whoistolu) [reply README.md]: - [ ] [OPTIONAL]
I want to acknowledge your hard work, @diarisdiakite  I have a couple of suggestions to enhance your presentation:

1. Including the mobile view of the presentation would have been exceptionally beneficial, especially since the project requirement emphasizes mobile responsiveness. The menu icon, Login, and some other elements appear a little disorganized because they were primarily styled for mobile view.

2. While it's commendable to communicate the effort invested in developing the app to your employer/client/reviewer, it might be more professional not to explicitly mention the presence of bugs. Instead, you could convey that the app is still a work in progress, leaving room for further refinement.

File: app/views/shared/_header_categories.html.erb (hunk 2 above)
Comment (Whoistolu) [reply app/views/shared/_header_categories.html.erb:28]: - [ ] [REQUIRED] In as much as I understand your explanation, I will suggest that you remove the gift credit that appears upon a user's Login to not confuse the user just as stated by the previous reviewer. This is not part of the project requirement. Let's leave the entire Category page to the control of the current logged-in user.


Screenshot 2023-12-02 at 19 07 40

File: app/views/categories/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        </div>
+      <% end %>
+    </ul>
+  </div> 
+
+  <div class="view-total-payment">
+    <%= "TOTAL PAYMENT" %>
+    $<%= @category.total_expense_amount %>
+  </div>
+</div>
```
Comment (Whoistolu) [inline app/views/categories/show.html.erb]: - [x] [REQUIRED] Kindly add a back button to this page so that the User will be able to navigate back to the  transactions page for that category as stated by the project requirement.

- [ ] [REQUIRED] As stated by the project requirement, For a given category, the list of transactions is presented ordered by the most recent.
-  At the top of the page the user should see the total amount for the category (sum of all of the amounts of transactions in this category2).
- [ ] There should be a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction which means the "Add a new expense" button on this page should be at the bottom of the screen.
- I have added three more expenses to this page but it is not showing up here. Kindly resolve these error as the first reviewer was unable o check them.

Screenshot 2023-12-02 at 19 58 37

Screenshot 2023-12-02 at 19 18 18

**Comment by diarisdiakite** [comment]
Hi @Whoistolu !
I can see that you added some comments, thank you, I am fixing those parts.
I have been waiting now for more than an hour (almost 2), and for some reasons I would have to drive home and it is getting late in my place. Would you please submit your review so that I can make all the changes?
Thank you in advance for your understanding. From home, I am unlikely to have electricity to complete the project.

File: app/views/categories/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<h1>New category</h1>
+
+<%= render "form_add", user: @user, category: @category %>
+
+<br>
```
Comment (Whoistolu) [inline app/views/categories/new.html.erb]: - [ ] [REQUIRED] I can see you are rendering the form that helps the user add a new category here, however, the project requirement states that there should be a back button that the user can use to navigate back to the home page.

Screenshot 2023-12-02 at 19 57 02

Screenshot 2023-12-02 at 19 55 03

File: README.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+- Create, Read and Delete expenses and categories
+- Authentication with Devise
+- Authorization rules for users
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+## 🚀 Live demo <a name="live-demo"></a>
+
+Please watch the presentation video at this [link](https://www.loom.com/share/678be0ed9a374054bc2f6ba576425e89)
```
Comment (Whoistolu) [inline README.md:66]: - [ ] [OPTIONAL]
I want to acknowledge your hard work, @diarisdiakite  I have a couple of suggestions to enhance your presentation:

1. Including the mobile view of the presentation would have been exceptionally beneficial, especially since the project requirement emphasizes mobile responsiveness. The menu icon, Login, and some other elements appear a little disorganized because they were primarily styled for mobile view.

2. While it's commendable to communicate the effort invested in developing the app to your employer/client/reviewer, it might be more professional not to explicitly mention the presence of bugs. Instead, you could convey that the app is still a work in progress, leaving room for further refinement.

File: app/views/shared/_header_categories.html.erb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        <li class="nav-item my-flex">
+          <%= link_to "Login", new_user_session_path, class:"button" %>
+          <%= link_to "Signup", new_user_registration_path, class:"button" %>
+        </li>
+      <% end %>
+  </ul>
+</nav>
+<div class="gift-credit-container details-heading">
+  <h2>Gift credit</h2>
+  <h2>$300</h2>
```
Comment (Whoistolu) [inline app/views/shared/_header_categories.html.erb:28]: - [ ] [REQUIRED] In as much as I understand your explanation, as recommended by the previous reviewer. I will suggest that you remove the gift credit that appears upon a user's Login to not confuse the user just as stated by the previous reviewer. This is not part of the project requirement. Let's leave the entire Category page to the control of the currently logged-in user.


Screenshot 2023-12-02 at 19 07 40

File: app/views/shared/_footer.html.erb (hunk 3 above)
Comment (Whoistolu) [reply app/views/shared/_footer.html.erb:3]: - Well resolved, however, let's remove the Gift Credit place holder added.
Screenshot 2023-12-02 at 19 14 14

File: db/migrate/20231122203839_add_name_to_categories.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class AddNameToCategories < ActiveRecord::Migration[7.0]
+  def change
+    add_column :categories, :name, :string
+  end
+end
```
Comment (Whoistolu) [inline db/migrate/20231122203839_add_name_to_categories.rb]: - [ ] [REQUIRED] While trying to fix the previous recommendations by the previous reviewer, there are new issues that have sprung up with the migration.
- The error message indicates that there is an attempt to add a column named "name" to the "categories" table in your database, but the column "name" already exists in that table. This results in a duplicate column error.

### To resolve this
- Manually delete this particular migration file and then run a 

``` ruby
rails db:rollback
```
Afterwards, you can try to migrate again.
Screenshot 2023-12-02 at 20 07 16

File: db/migrate/20231124002142_remove_icon_from_categories.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class RemoveIconFromCategories < ActiveRecord::Migration[7.0]
+  def change
+    remove_column :categories, :icon, :string
+  end
+end
```
Comment (Whoistolu) [inline db/migrate/20231124002142_remove_icon_from_categories.rb]: - [ ] [REQUIRED] There is another migration error here @diarisdiakite and the error indicates that there is an attempt to remove a column named "icon" from the "categories" table in your database, but the column "icon" does not exist in that table. This results in an undefined column error.

Screenshot 2023-12-02 at 20 13 19

### To resolve this error
- Kindly remove this migration file manually and run a rollback as suggested earlier.  Then run a migration. That was how I was able to get the application working.

**Review by Whoistolu — CHANGES_REQUESTED** [review]
Hi @diarisdiakite 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [ ]  _Check the inline comments under the review for sections to improve._

### Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/categories/show.html.erb (hunk 5 above)
Comment (diarisdiakite) [reply app/views/categories/show.html.erb]: I don't understand @Whoistolu. Please telle me where should the total come... 

image

Comment (Whoistolu) [reply app/views/categories/show.html.erb]: Apologies for any inconvenience, @diarisdiakite. Here are the project requirements:

1. Please ensure that the user can view the total amount of all transactions for each category at the top of the page.
2. Additionally, make sure that the list of transactions is displayed in descending order, with the most recent transactions appearing first.
3. Lastly, move the "Add a new expense" button to the bottom of the screen.

File: app/views/shared/_header_categories.html.erb (hunk 2 above)
Comment (diarisdiakite) [reply app/views/shared/_header_categories.html.erb]: For the credit gift I added a conditional sign_in to display it

image


But @Whoistolu please confirm if contrary to the design it has to be completely removed. As discussed, I followed the design.

File: app/views/categories/show.html.erb
Comment (fobadara) [inline app/views/categories/show.html.erb:1]: [REQUIRED] I can't see the back button requested by the previous reviewer. Kindly include it

[REQUIRED] As suggested by the previous reviewer kindly order the transaction by the most recent

[REQUIRED] As suggested by the previous reviewer kindly ensure that the user can view the total of all transactions for each category at the top of the page

File: app/views/categories/new.html.erb (hunk 6 above)
Comment (fobadara) [inline app/views/categories/new.html.erb]: [REQUIRED] Please include the back button as stated by the previous reviewer

**Review by fobadara — CHANGES_REQUESTED** [review]
Hi @diarisdiakite, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which haven't been addressed yet.

REDO

## Highlights
✅ Descriptive README
✅ Descriptive PR summary
✅ Clean code

 ## Required Changes ♻️
_Check the comments under the review._
I had to include some comments here because it could not be done through inline comments
[REQUIRED] As stated by the previous reviewer kindly fix the migration errors in the images below:
The first error
image

The second error
image


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/categories/show.html.erb
Comment (diarisdiakite) [reply app/views/categories/show.html.erb:1]: Done @fobadara 
For the migration, I am confused, as I can see all the status up
image

File: app/views/categories/new.html.erb (hunk 6 above)
Comment (diarisdiakite) [reply app/views/categories/new.html.erb]: Done @fobadara

File: app/assets/stylesheets/application.css [hunk 11]
```diff
@@ -128,6 +128,7 @@ li {
   width: 100%;
   min-height: 100vh;
   display: flex;
+  overflow: scroll;
 }
```
Comment (zieeco) [inline app/assets/stylesheets/application.css:133]: *NOTE: I'm using this inline feature to add an inline comment. This comment is not related to this code here*

- Hey, to fix the migration issue, delete these two migration files to resolve the issue

image

image

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @diarisdiakite,




Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the changes, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- Great effort making the requested changes.✔️
- Nice design. ✔️
- Descriptive PR. ✔️


## Required Changes ♻️
_Check the comments under the review._





## Optional suggestions

- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/assets/stylesheets/application.css (hunk 11 above)
Comment (diarisdiakite) [reply app/assets/stylesheets/application.css:133]: Done @zieeco. Thank you for the suggestion. I really appreciate.

**Review by zieeco — APPROVED** [review]
# status: APPROVED :green_circle: :100: 

Hi, @diarisdiakite, 


68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f78543078657a514755357843444a754350652f67697068792e676966


Good job so far 👍 Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights ⚡ 

✅ Descriptive PR 👍 
✅ Nice design 🔥 
✅ Requested changes implemented 🔥 


## Optional suggestions

- N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
